cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/plugin/PluginInstantiator.java

```diff
@@ -211,22 +220,57 @@ public ClassLoader getArtifactClassLoader(ArtifactId artifactId) throws IOExcept
     }
   }
 
-  private PluginProperties substituteMacroFields(Plugin plugin, MacroEvaluator macroEvaluator) {
+  private PluginProperties substituteMacros(Plugin plugin, @Nullable MacroEvaluator macroEvaluator) {
     Map<String, String> properties = new HashMap<>();
     Map<String, PluginPropertyField> pluginPropertyFieldMap = plugin.getPluginClass().getProperties();
-    MacroParser macroParser = new MacroParser(macroEvaluator);
+
+    // create macro evaluator and parser based on if it is config or runtime
+    boolean configTime = macroEvaluator == null;
+    TrackingMacroEvaluator trackingMacroEvaluator = new TrackingMacroEvaluator();
+    MacroParser macroParser;
+    if (configTime) {
+      macroParser = new MacroParser(trackingMacroEvaluator);
+    } else {
+      macroParser = new MacroParser(macroEvaluator);
+    }
 
     for (Map.Entry<String, String> property : plugin.getProperties().getProperties().entrySet()) {
       PluginPropertyField field = pluginPropertyFieldMap.get(property.getKey());
+      String propertyValue = property.getValue();
       if (field != null && field.isMacroSupported()) {
-        properties.put(property.getKey(), macroParser.parse(property.getValue()));
-      } else {
-        properties.put(property.getKey(), property.getValue());
+        if (configTime) {
+          // parse for syntax check and check if trackingMacroEvaluator finds macro syntax present
+          macroParser.parse(propertyValue);
```

Comment:
this seems like it can be cleaned up a bit after https://github.com/caskdata/cdap/pull/6089 is merged, as the Plugin class will store what macros exist, so we won't have to parse again.

Lets comment a todo and make sure that PR cleans this up.